### PR TITLE
bug: Unable to Change CharacterComponent's collider shape properties without causing crash

### PR DESCRIPTION
**Pull request by dloe**
# PR Details

Assigning a new `ColliderShape` into `CharacterComponent` and running `ComposeShape()` causes a UAF related exception to be thrown. Related to an issue with the `KinematicCharacter` reference inside of the `discreteDynamicWorld` in the `Simulation` object. 

Solution would be to add additional setup to a overridden `ComposeShape()` method inside CharacterComponent. This setup would include the complete removal, reconstruction and reintegration of the KinematicCharacterController to the Simulation. Also added an additional function to try and maintain various physics properties onto the `KinematicCharacter` such as Gravity, Fallspeed, etc. 

## Related Issue

https://github.com/stride3d/stride/issues/2005

## Types of changes

- [ ] Docs change / refactoring / dependency upgrade
- [X] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

- [ ] My change requires a change to the documentation.
- [ ] I have added tests to cover my changes.
- [X] All new and existing tests passed.
- [X] **I have built and run the editor to try this change out.**

### CODE REVIEW

File: sources/engine/Stride.Physics/Elements/CharacterComponent.cs [hunk 1]
```diff
@@ -300,6 +300,54 @@ protected override void OnAttach()
...
+                KinematicCharacter.Dispose();
+                KinematicCharacter = null;
+                BulletSharp.Math.Vector3 unitY = new BulletSharp.Math.Vector3(0f, 1f, 0f);
+                //Make new KinematicCharacter
+                KinematicCharacter = new BulletSharp.KinematicCharacterController((BulletSharp.PairCachingGhostObject)NativeCollisionObject, (BulletSharp.ConvexShape)ColliderShape.InternalShape, StepHeight, ref unitY);
+                OverrideKinematicCharacterValues(kinematicCharacterProperties);
+
+                //now we can add these references BACK in Simulation's discreteDynamicsWorld
+                Simulation.AddCharacter(this, (CollisionFilterGroupFlags)CollisionGroup, CanCollideWith);
+            }
```
Comment (Eideren) [inline sources/engine/Stride.Physics/Elements/CharacterComponent.cs:335]: It seems like virtualizing `ComposeShape` may not be necessary as that method runs right before `OnAttach`, 
https://github.com/stride3d/stride/blob/4ebbf54839cf762ef8e9d135c78f453cf30a723f/sources/engine/Stride.Physics/Engine/PhysicsComponent.cs#L660-L672
So, it you could move this section inside `OnAttach` instead and have it work the same way on top of being able to reduce logic duplication ? I might have missed something though ... let me know !

**Review by Eideren — COMMENTED** [review]
Looks pretty good, just a small question

File: sources/engine/Stride.Physics/Elements/CharacterComponent.cs (hunk 1 above)
Comment (dloe) [reply sources/engine/Stride.Physics/Elements/CharacterComponent.cs:335]: That is a good thing to point out. However due to the fact that users running `ComposeShape` directly in their projects for this scenario can cause the crash, I ended up using the virtual `ComposeShape` approach. Sorry for that delay, I wanted to run some more tests just to be sure and wasnt at my normal computer.

Comment (Eideren) [reply sources/engine/Stride.Physics/Elements/CharacterComponent.cs:335]: Makes sense, good thinking - we're all working on our free time, don't worry about the delay !

**Comment by Eideren** [comment]
Thanks !

**Pull request merged**